backend/code_coverage_backend/gcp.py

```diff
@@ -85,108 +75,115 @@ def __init__(self, reports_dir=None):
 
-    def ingest_pushes(self, repository, min_push_id=None, nb_pages=3):
+    def ingest_pushes(self, repository, platform, suite, min_push_id=None, nb_pages=3):
         """
         Ingest HGMO changesets and pushes into our Redis Cache
         The pagination goes from oldest to newest, starting from the optional min_push_id
         """
-        chunk_size = 8
-        params = {"version": 2}
-        if min_push_id is not None:
-            assert isinstance(min_push_id, int)
-            params["startID"] = min_push_id
-            params["endID"] = min_push_id + chunk_size
-
-        for page in range(nb_pages):
-
-            r = requests.get(
-                HGMO_PUSHES_URL.format(repository=repository), params=params
-            )
-            data = r.json()
-
-            # Sort pushes to go from oldest to newest
-            pushes = sorted(
-                [(int(push_id), push) for push_id, push in data["pushes"].items()],
-                key=lambda p: p[0],
-            )
-            if not pushes:
-                return
-
-            for push_id, push in pushes:
-
-                changesets = push["changesets"]
-                date = push["date"]
-                self.store_push(repository, push_id, changesets, date)
-
-                reports = [
-                    changeset
-                    for changeset in changesets
-                    if self.ingest_report(repository, push_id, changeset, date)
-                ]
-                if reports:
-                    logger.info("Found reports in that push", push_id=push_id)
-
-            newest = pushes[-1][0]
-            params["startID"] = newest
-            params["endID"] = newest + chunk_size
-
-    def ingest_report(self, repository, push_id, changeset, date):
+        ingested = False
+        for push_id, push in hgmo_pushes(repository, min_push_id, nb_pages):
+            for changeset in push["changesets"]:
+                # TODO: look all neighboring reports on GCP
+                report = Report(
+                    self.reports_dir,
+                    repository,
+                    changeset,
+                    platform,
+                    suite,
+                    push_id=push_id,
+                    date=push["date"],
+                )
+
+                # Always link changeset to push to find closest available report
+                self.redis.hmset(
+                    KEY_CHANGESET.format(
+                        repository=report.repository, changeset=report.changeset
+                    ),
+                    {"push": report.push_id, "date": report.date},
+                )
+
+                if not ingested and self.ingest_report(report):
+                    logger.info(
+                        "Found report in that push", push_id=push_id, report=str(report)
+                    )
+
+                    # Only ingest first report found in a push in order to stay below 30s response time
```

Comment:
If we don't ingest them, what happens with the history?

Reply:
I think there is rarely (if ever) multiple reports in a single push, so that should not happen frequently.

A followup i have in mind to speed up that part would be to:
1. list all the available revisions on the bucket ( a bit like in `ingest_available_reports`) : it's prett. fast.
2. use that list to check the presence of a specific report instead of doing a `HEAD` request for each changeset in a push.

Reply:
SGTM, can you file a follow-up issue so we don't forget?

Reply:
https://github.com/mozilla/code-coverage/issues/149